crates/intermediate-theorem-format/src/data/types.rs

```diff
@@ -0,0 +1,283 @@
+                    .unwrap_or_else(|| {
+                        panic!(
+                            "BUG: Datatype must be registered in struct_id_map before type translation. \
+                             QualifiedId: {:?}, Total registered datatypes: {}",
+                            qualified_id,
+                            struct_id_map.len()
+                        )
+                    });
+
+                // Recursively resolve type arguments
+                let resolved_type_args: Vec<TheoremType> = type_args.iter()
+                    .map(|arg| Self::from_move_type(arg, env, struct_id_map))
+                    .collect();
+
+                TheoremType::Struct {
+                    struct_id: theorem_struct_id,
+                    type_args: resolved_type_args,
+                }
+            }
+            Type::Vector(elem_ty) => {
+                TheoremType::Vector(Box::new(Self::from_move_type(elem_ty, env, struct_id_map)))
+            }
+            Type::Reference(_, ty) => {
+                TheoremType::Reference(Box::new(Self::from_move_type(ty, env, struct_id_map)))
+            }
+            Type::TypeParameter(idx) => TheoremType::TypeParameter(*idx),
+            Type::Tuple(tys) => {
+                TheoremType::Tuple(tys.iter().map(|t| Self::from_move_type(t, env, struct_id_map)).collect())
+            }
+            _ => unreachable!(),
+        }
+    }
+
+    /// Wrap this type in ProgramState monad
+    pub fn wrap_in_monad(self) -> Self {
+        TheoremType::ProgramState(Box::new(self))
+    }
+
+    /// Check if this is a ProgramState type
+    pub fn is_monad(&self) -> bool {
+        matches!(self, TheoremType::ProgramState(_))
+    }
+
+    /// Unwrap the inner type from ProgramState, if applicable
+    pub fn unwrap_monad(&self) -> Option<&TheoremType> {
+        match self {
+            TheoremType::ProgramState(inner) => Some(inner),
+            _ => None,
+        }
+    }
+
+    /// Sanitize a name for use in Lean identifiers
+    pub fn sanitize_name(name: &str) -> String {
+        name.replace("::", "_")
+            .replace("<", "")
+            .replace(">", "")
+            .replace("$", "_")
+            .replace("%", "_")
+            .replace("#", "_")
+            .replace("@", "_")
+            .chars()
+            .filter(|c| c.is_alphanumeric() || *c == '_')
+            .collect()
+    }
+
+    // ========================================================================
+    // Functional Traversal API (Visitor Pattern)
+    // ========================================================================
+
+    /// Fold over all types in the type tree (pre-order traversal)
+    /// The closure receives the accumulator and a reference to each type
+    pub fn fold<T>(&self, init: T, mut f: impl FnMut(T, &TheoremType) -> T) -> T {
+        self.fold_impl(init, &mut f)
+    }
+
+    fn fold_impl<T>(&self, acc: T, f: &mut impl FnMut(T, &TheoremType) -> T) -> T {
```

Comment:
is it possible to implement a generic iterator, and then `fold`/`all`/`any` would be just the standard operations on iterators?

Reply:
Implemented!